File: src/main/java/net/jgp/books/spark/ch08/lab400_elasticsearch_Ingestion/ElasticsearchToDatasetApp.java
package net.jgp.books.spark.ch08.lab400_elasticsearch_Ingestion;

import org.apache.spark.sql.Dataset;
import org.apache.spark.sql.Row;
import org.apache.spark.sql.SparkSession;

/**
 * Connects Spark and Elasticsearch, ingests data from Elasticsearch in
 * Spark.
 * 
 * @author jgp
 */
public class ElasticsearchToDatasetApp {

  /**
   * Our main...
   * 
   * @param args
   */
  public static void main(String[] args) {
    ElasticsearchToDatasetApp app =
        new ElasticsearchToDatasetApp();
    app.start();
  }

  /**
   * The real processing starts.
   */
  private void start() {
    long t0 = System.currentTimeMillis();

    SparkSession spark = SparkSession.builder()
        .appName("Elasticsearch to Dataframe")
        .master("local")
        .getOrCreate();

    long t1 = System.currentTimeMillis();
    System.out.println("Getting a session took: " + (t1 - t0) + " ms");

    Dataset<Row> df = spark
        .read()
        .format("org.elasticsearch.spark.sql")
        .option("es.nodes", "localhost")
        .option("es.port", "9200")
        .option("es.query", "?q=*")
        .option("es.read.field.as.array.include", "Inspection_Date")
        .load("nyc_restaurants");

    long t2 = System.currentTimeMillis();
    System.out.println(
        "Init communication and starting to get some results took: "
            + (t2 - t1) + " ms");

    // Shows only a few records as they are pretty long
    df.show(10);

    long t3 = System.currentTimeMillis();
    System.out.println("Showing a few records took: " + (t3 - t2) + " ms");

    df.printSchema();
    long t4 = System.currentTimeMillis();
    System.out.println("Displaying the schema took: " + (t4 - t3) + " ms");

    System.out.println("The dataframe contains " +
        df.count() + " record(s).");
    long t5 = System.currentTimeMillis();
    System.out.println("Counting the number of records took: " + (t5 - t4)
        + " ms");

    System.out.println("The dataframe is split over " + df.rdd()
        .getPartitions().length + " partition(s).");
    long t6 = System.currentTimeMillis();
    System.out.println("Counting the # of partitions took: " + (t6 - t5)
        + " ms");
  }
}

File: src/main/python/lab400_elasticsearch_Ingestion/elasticsearchToDatasetApp.py
"""
 Connects PySpark and Elasticsearch, ingests data from Elasticsearch in PySpark.

 @author rambabu.posa
"""
import time
import logging
from pyspark.sql import SparkSession

t0 = int(round(time.time() * 1000))

# Creates a session on a local master
spark = SparkSession.builder.appName("Elasticsearch to Dataframe") \
    .master("local[*]").getOrCreate()

t1 = int(round(time.time() * 1000))

logging.info("Getting a session took: {} ms".format((t1 - t0)))

df = spark.read.format("org.elasticsearch.spark.sql") \
    .option("es.nodes", "localhost") \
    .option("es.port", "9200") \
    .option("es.query", "?q=*") \
    .option("es.read.field.as.array.include", "Inspection_Date") \
    .load("nyc_restaurants")

t2 = int(round(time.time() * 1000))

logging.info("Init communication and starting to get some results took: {} ms".format((t2 - t1)))

# Shows only a few records as they are pretty long
df.show(10)

t3 = int(round(time.time() * 1000))

logging.info("Showing a few records took: {} ms".format((t3 - t2)))

df.printSchema()
t4 = int(round(time.time() * 1000))

logging.info("Displaying the schema took: {} ms".format((t4 - t3)))

logging.info("The dataframe contains ${df.count} record(s).")

t5 = int(round(time.time() * 1000))
logging.info("Counting the number of records took: {} ms".format((t5 - t4)))

logging.info("The dataframe is split over {} partition(s).".format(df.rdd.getNumPartitions()))

t6 = int(round(time.time() * 1000))

logging.info("Counting the # of partitions took: {} ms".format((t6 - t5)))

spark.stop()